File: core/src/ten_runtime/binding/python/interface/ten/ten_env.py
#
# Copyright © 2025 Agora
# This file is part of TEN Framework, an open source project.
# Licensed under the Apache License, Version 2.0, with certain conditions.
# Refer to the "LICENSE" file in the root directory for more information.
#
from typing import Callable, Optional

from libten_runtime_python import _Extension, _TenEnv

from .error import TenError
from .cmd_result import CmdResult
from .cmd import Cmd
from .video_frame import VideoFrame
from .audio_frame import AudioFrame
from .data import Data
from .ten_env_base import TenEnvBase


ResultHandler = Callable[
    ["TenEnv", Optional[CmdResult], Optional[TenError]], None
]

ErrorHandler = Callable[["TenEnv", Optional[TenError]], None]


class TenEnv(TenEnvBase):

    def __init__(self, internal_obj: _TenEnv) -> None:
        super().__init__(internal_obj)
        self._release_handler = None

    def __del__(self) -> None:
        pass

    def _set_release_handler(self, handler: Callable[[], None]) -> None:
        self._release_handler = handler

    def _on_release(self) -> None:
        if self._release_handler is not None:
            self._release_handler()

    def on_configure_done(self) -> None:
        return self._internal.on_configure_done()

    def on_init_done(self) -> None:
        return self._internal.on_init_done()

    def on_start_done(self) -> None:
        return self._internal.on_start_done()

    def on_stop_done(self) -> None:
        return self._internal.on_stop_done()

    def on_deinit_done(self) -> None:
        return self._internal.on_deinit_done()

    def on_create_instance_done(self, instance: _Extension, context) -> None:
        return self._internal.on_create_instance_done(instance, context)

    def get_property_to_json(
        self, path: Optional[str] = None
    ) -> tuple[str, Optional[TenError]]:
        return self._internal.get_property_to_json(path)

    def set_property_from_json(
        self, path: str, json_str: str
    ) -> Optional[TenError]:
        return self._internal.set_property_from_json(path, json_str)

    def send_cmd(
        self, cmd: Cmd, result_handler: Optional[ResultHandler] = None
    ) -> Optional[TenError]:
        return self._internal.send_cmd(cmd, result_handler, False)

    def send_cmd_ex(
        self, cmd: Cmd, result_handler: Optional[ResultHandler] = None
    ) -> Optional[TenError]:
        return self._internal.send_cmd(cmd, result_handler, True)

    def send_data(
        self, data: Data, error_handler: Optional[ErrorHandler] = None
    ) -> Optional[TenError]:
        return self._internal.send_data(data, error_handler)

    def send_video_frame(
        self,
        video_frame: VideoFrame,
        error_handler: Optional[ErrorHandler] = None,
    ) -> Optional[TenError]:
        return self._internal.send_video_frame(video_frame, error_handler)

    def send_audio_frame(
        self,
        audio_frame: AudioFrame,
        error_handler: Optional[ErrorHandler] = None,
    ) -> Optional[TenError]:
        return self._internal.send_audio_frame(audio_frame, error_handler)

    def return_result(
        self,
        result: CmdResult,
        error_handler: Optional[ErrorHandler] = None,
    ) -> Optional[TenError]:
        return self._internal.return_result(result, error_handler)

    def is_property_exist(self, path: str) -> tuple[bool, Optional[TenError]]:
        return self._internal.is_property_exist(path)

    def get_property_int(self, path: str) -> tuple[int, Optional[TenError]]:
        return self._internal.get_property_int(path)

    def set_property_int(self, path: str, value: int) -> Optional[TenError]:
        return self._internal.set_property_int(path, value)

    def get_property_string(self, path: str) -> tuple[str, Optional[TenError]]:
        return self._internal.get_property_string(path)

    def set_property_string(self, path: str, value: str) -> Optional[TenError]:
        return self._internal.set_property_string(path, value)

    def get_property_bool(self, path: str) -> tuple[bool, Optional[TenError]]:
        return self._internal.get_property_bool(path)

    def set_property_bool(self, path: str, value: bool) -> Optional[TenError]:
        if value:
            return self._internal.set_property_bool(path, 1)
        else:
            return self._internal.set_property_bool(path, 0)

    def get_property_float(self, path: str) -> tuple[float, Optional[TenError]]:
        return self._internal.get_property_float(path)

    def set_property_float(self, path: str, value: float) -> Optional[TenError]:
        return self._internal.set_property_float(path, value)

    def init_property_from_json(self, json_str: str) -> Optional[TenError]:
        return self._internal.init_property_from_json(json_str)

File: core/src/ten_runtime/binding/nodejs/interface/ten_env/ten_env.ts
//
// Copyright © 2025 Agora
// This file is part of TEN Framework, an open source project.
// Licensed under the Apache License, Version 2.0, with certain conditions.
// Refer to the "LICENSE" file in the root directory for more information.
//
import { AudioFrame } from "../msg/audio_frame";
import { VideoFrame } from "../msg/video_frame";
import { Cmd } from "../msg/cmd";
import { CmdResult } from "../msg/cmd_result";
import { Data } from "../msg/data";
import ten_addon from "../ten_addon";
import { LogLevel } from "./log_level";

export class TenEnv {
  async sendCmd(cmd: Cmd): Promise<[CmdResult | null, Error | null]> {
    return new Promise<[CmdResult | null, Error | null]>((resolve) => {
      const err = ten_addon.ten_nodejs_ten_env_send_cmd(
        this,
        cmd,
        async (cmdResult: CmdResult | null, error: Error | null) => {
          resolve([cmdResult, error]);
        }
      );

      if (err) {
        resolve([null, err]);
      }
    });
  }

  async sendData(data: Data): Promise<Error | null> {
    return new Promise<Error | null>((resolve) => {
      const err = ten_addon.ten_nodejs_ten_env_send_data(
        this,
        data,
        async (error: Error | null) => {
          resolve(error);
        }
      );

      if (err) {
        resolve(err);
      }
    });
  }

  async sendVideoFrame(videoFrame: VideoFrame): Promise<Error | null> {
    return new Promise<Error | null>((resolve) => {
      const err = ten_addon.ten_nodejs_ten_env_send_video_frame(
        this,
        videoFrame,
        async (error: Error | null) => {
          resolve(error);
        }
      );

      if (err) {
        resolve(err);
      }
    });
  }

  async sendAudioFrame(audioFrame: AudioFrame): Promise<Error | null> {
    return new Promise<Error | null>((resolve) => {
      const err = ten_addon.ten_nodejs_ten_env_send_audio_frame(
        this,
        audioFrame,
        async (error: Error | null) => {
          resolve(error);
        }
      );

      if (err) {
        resolve(err);
      }
    });
  }

  async returnResult(cmdResult: CmdResult): Promise<Error | null> {
    return new Promise<Error | null>((resolve) => {
      const err = ten_addon.ten_nodejs_ten_env_return_result(
        this,
        cmdResult,
        async (error: Error | null) => {
          resolve(error);
        }
      );

      if (err) {
        resolve(err);
      }
    });
  }

  async isPropertyExist(path: string): Promise<boolean> {
    return new Promise<boolean>((resolve) => {
      ten_addon.ten_nodejs_ten_env_is_property_exist(
        this,
        path,
        async (result: boolean) => {
          resolve(result);
        }
      );
    });
  }

  async getPropertyToJson(path: string): Promise<[string, Error | null]> {
    return new Promise<[string, Error | null]>((resolve) => {
      const err = ten_addon.ten_nodejs_ten_env_get_property_to_json(
        this,
        path,
        async (result: string, error: Error | null) => {
          resolve([result, error]);
        }
      );

      if (err) {
        resolve(["", err]);
      }
    });
  }

  async setPropertyFromJson(path: string, jsonStr: string): Promise<Error | null> {
    return new Promise<Error | null>((resolve) => {
      const err = ten_addon.ten_nodejs_ten_env_set_property_from_json(
        this,
        path,
        jsonStr,
        async (error: Error | null) => {
          resolve(error);
        }
      );

      if (err) {
        resolve(err);
      }
    });
  }

  async getPropertyNumber(path: string): Promise<[number, Error | null]> {
    return new Promise<[number, Error | null]>((resolve) => {
      const err = ten_addon.ten_nodejs_ten_env_get_property_number(
        this,
        path,
        async (result: number, error: Error | null) => {
          resolve([result, error]);
        }
      );

      if (err) {
        resolve([0, err]);
      }
    });
  }

  async setPropertyNumber(path: string, value: number): Promise<Error | null> {
    return new Promise<Error | null>((resolve) => {
      const err = ten_addon.ten_nodejs_ten_env_set_property_number(
        this,
        path,
        value,
        async (error: Error | null) => {
          resolve(error);
        }
      );

      if (err) {
        resolve(err);
      }
    });
  }

  async getPropertyString(path: string): Promise<[string, Error | null]> {
    return new Promise<[string, Error | null]>((resolve) => {
      const err = ten_addon.ten_nodejs_ten_env_get_property_string(
        this,
        path,
        async (result: string, error: Error | null) => {
          resolve([result, error]);
        }
      );

      if (err) {
        resolve(["", err]);
      }
    });
  }

  async setPropertyString(path: string, value: string): Promise<Error | null> {
    return new Promise<Error | null>((resolve) => {
      const err = ten_addon.ten_nodejs_ten_env_set_property_string(
        this,
        path,
        value,
        async (error: Error | null) => {
          resolve(error);
        }
      );

      if (err) {
        resolve(err);
      }
    });
  }

  async initPropertyFromJson(jsonStr: string): Promise<Error | null> {
    return new Promise<Error | null>((resolve) => {
      const err = ten_addon.ten_nodejs_ten_env_init_property_from_json(
        this,
        jsonStr,
        async (error: Error | null) => {
          resolve(error);
        }
      );

      if (err) {
        resolve(err);
      }
    });
  }

  logVerbose(message: string): Error | null {
    return this.log_internal(LogLevel.VERBOSE, message);
  }

  logDebug(message: string): Error | null {
    return this.log_internal(LogLevel.DEBUG, message);
  }

  logInfo(message: string): Error | null {
    return this.log_internal(LogLevel.INFO, message);
  }

  logWarn(message: string): Error | null {
    return this.log_internal(LogLevel.WARN, message);
  }

  logError(message: string): Error | null {
    return this.log_internal(LogLevel.ERROR, message);
  }

  logFatal(message: string): Error | null {
    return this.log_internal(LogLevel.FATAL, message);
  }

  private log_internal(level: number, message: string): Error | null {
    const _prepareStackTrace = Error.prepareStackTrace;
    Error.prepareStackTrace = (_, stack): NodeJS.CallSite[] => stack;
    const stack_ = new Error().stack as unknown as NodeJS.CallSite[];
    const stack = stack_.slice(1);
    Error.prepareStackTrace = _prepareStackTrace;

    const _callerFile = stack[1].getFileName();
    const _callerLine = stack[1].getLineNumber();
    const _callerFunction = stack[1].getFunctionName();

    const callerFile = _callerFile ? _callerFile : "unknown";
    const callerLine = _callerLine ? _callerLine : 0;
    const callerFunction = _callerFunction ? _callerFunction : "anonymous";

    return ten_addon.ten_nodejs_ten_env_log_internal(
      this,
      level,
      callerFunction,
      callerFile,
      callerLine,
      message
    );
  }
}

ten_addon.ten_nodejs_ten_env_register_class(TenEnv);
